### PR TITLE
Limit max reset frames to mitigate HTTP/2 RST floods

### PR DESCRIPTION
**Pull request by ikhoon**
Motivation:

To mitigate against the "HTTP/2 Rapid Reset" attack, it is recommended that HTTP/2 servers should close connections that exceed the concurrent stream limit.

Reference:

- https://blog.cloudflare.com/technical-breakdown-http2-rapid-reset-ddos-attack/
- https://www.cve.org/CVERecord?id=CVE-2023-44487
- https://github.com/netty/netty/commit/58f75f665aa81a8cbcf6ffa74820042a285c5e61#diff-82f568a075ff63e9727ce8622f3a2b1553099182edf1fd0b4f857226252b05adR47

Modifications:

- Add `ServerBuilder.http2MaxRestFramesPerWindow()` option and `-Dcom.linecorp.armeria.defaultHttp2MaxResetFramesPerMinute<integer>` property to limit the maximum allowed RST frames.
  - If not set, 400 RST frames per minute are allowed by default.
- Bump Netty version to 4.1.100 from 4.1.96

Result:

You can now protect your server against DDOS caused by RST floods.
```java
Server
  .builder()
  .http2MaxResetFramesPerWindow(100, 10)
  .build();
```

### CODE REVIEW

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 1]
```diff
@@ -1032,6 +1036,22 @@ public static long defaultHttp2MaxHeaderListSize() {
         return DEFAULT_HTTP2_MAX_HEADER_LIST_SIZE;
     }
 
+    /**
+     * Returns the default maximum number RST frames that are allowed per window before the connection is
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java:1040]: nit: number of

File: core/src/main/java/com/linecorp/armeria/common/FlagsProvider.java [hunk 2]
```diff
@@ -694,6 +694,22 @@ default Long defaultHttp2MaxHeaderListSize() {
         return null;
     }
 
+    /**
+     * Returns the default maximum number RST frames that are allowed per window before the connection is
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/common/FlagsProvider.java:698]: ditto

**Review by minwoox — APPROVED** [review]
Looks great!

File: core/src/main/java/com/linecorp/armeria/common/Flags.java [hunk 3]
```diff
@@ -1032,6 +1036,22 @@ public static long defaultHttp2MaxHeaderListSize() {
         return DEFAULT_HTTP2_MAX_HEADER_LIST_SIZE;
     }
 
+    /**
+     * Returns the default maximum number of RST frames that are allowed per window before the connection is
+     * closed. This allows to protect against the remote peer flooding us with such frames and so use up a lot
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/common/Flags.java:1041]:
```suggestion
     * closed. This allows to protect against the remote peer flooding us with such frames and using up a lot
```

File: core/src/main/java/com/linecorp/armeria/common/FlagsProvider.java [hunk 4]
```diff
@@ -694,6 +694,22 @@ default Long defaultHttp2MaxHeaderListSize() {
         return null;
     }
 
+    /**
+     * Returns the default maximum number of RST frames that are allowed per window before the connection is
+     * closed. This allows to protect against the remote peer flooding us with such frames and so use up a lot
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/common/FlagsProvider.java:699]:
```suggestion
     * closed. This allows to protect against the remote peer flooding us with such frames and using up a lot
```

File: core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java [hunk 5]
```diff
@@ -770,6 +772,26 @@ public ServerBuilder http2MaxStreamsPerConnection(long http2MaxStreamsPerConnect
         return this;
     }
 
+    /**
+     * Sets the maximum number of RST frames that are allowed per window before the connection is closed. This
+     * allows to protect against the remote peer flooding us with such frames and so use up a lot of CPU.
```
Comment (jrhee17) [inline core/src/main/java/com/linecorp/armeria/server/ServerBuilder.java:777]:
```suggestion
     * allows to protect against the remote peer flooding us with such frames and using up a lot of CPU.
```

**Review by jrhee17 — APPROVED** [review]
Thanks for the quick fix @ikhoon ! 🙇 👍 🙇

File: core/src/main/java/com/linecorp/armeria/server/ServerConfig.java [hunk 6]
```diff
@@ -213,6 +213,19 @@ public interface ServerConfig {
      */
     long http2MaxHeaderListSize();
 
+    /**
+     * Returns the maximum number of RST frames that are allowed per {@link #http2MaxResetFramesWindowSeconds()}.
```
Comment (minwoox) [inline core/src/main/java/com/linecorp/armeria/server/ServerConfig.java:217]:
```suggestion
     * Returns the maximum number of RST frames that are allowed per
     * {@link #http2MaxResetFramesWindowSeconds()}.
```

**Comment by minwoox** [comment]
👍 👍 👍

**Pull request merged**